[lake/tiering] Tiering job leaks lake writers on failure and masks the real exception

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.


### Fluss version

0.9.0 (latest release)

### Please describe the bug 🐞

When a tiering task fails and Flink triggers a failover, the close path throws "Memory was leaked by query", masking the real failure:
```
2026-07-21 12:04:14,661 INFO  org.apache.flink.connector.base.source.reader.SourceReaderBase [] - Closing Source Reader.
2026-07-21 12:04:14,661 INFO  org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher [] - Shutting down split fetcher 42
2026-07-21 12:04:14,819 INFO  org.apache.fluss.client.table.scanner.log.RemoteLogDownloader$DownloadRemoteLogThread [] - Shutting down
2026-07-21 12:04:14,819 INFO  org.apache.fluss.client.table.scanner.log.RemoteLogDownloader$DownloadRemoteLogThread [] - Stopped
2026-07-21 12:04:14,820 INFO  org.apache.fluss.client.table.scanner.log.RemoteLogDownloader$DownloadRemoteLogThread [] - Shutdown completed
2026-07-21 12:04:14,868 ERROR org.apache.arrow.memory.BaseAllocator                        [] - Memory was leaked by query. Memory leaked: (81920)
Outstanding child allocators : 
  Allocator(system-column-allocator) 0/81920/81920/9223372036854775807 (res/actual/peak/limit)
Allocator(ROOT) 0/81920/4742874/9223372036854775807 (res/actual/peak/limit)

2026-07-21 12:04:14,869 ERROR org.apache.flink.connector.base.source.reader.fetcher.SplitFetcherManager [] - Received uncaught exception.
java.lang.RuntimeException: Failed to close Arrow buffer allocator. Arrow batches returned by pollRecordBatch() must be closed before closing the scanner.
	at org.apache.fluss.record.LogRecordReadContext.close(LogRecordReadContext.java:401) ~[xxx]
	at org.apache.fluss.client.table.scanner.log.LogFetcher.close(LogFetcher.java:617) ~[xxx]
	at org.apache.fluss.client.table.scanner.log.LogScannerImpl.close(LogScannerImpl.java:375) ~[xxx]
	at org.apache.fluss.flink.tiering.source.TieringSplitReader.close(TieringSplitReader.java:768) ~[xxx]
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.run(SplitFetcher.java:124) [xxx]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:991) [?:?]
Caused by: java.lang.IllegalStateException: Memory was leaked by query. Memory leaked: (81920)
Outstanding child allocators : 
  Allocator(system-column-allocator) 0/81920/81920/9223372036854775807 (res/actual/peak/limit)
Allocator(ROOT) 0/81920/4742874/9223372036854775807 (res/actual/peak/limit)

	at org.apache.arrow.memory.BaseAllocator.close(BaseAllocator.java:504) ~[blob_p-5c46eb4a000924a0af48b02fdbab6d18c382ee46-3b580ff3b0d21175f35f1083631c63c1:xxx]
	at org.apache.arrow.memory.RootAllocator.close(RootAllocator.java:27) ~[blob_p-5c46eb4a000924a0af48b02fdbab6d18c382ee46-3b580ff3b0d21175f35f1083631c63c1:xxx]
	at org.apache.fluss.record.LogRecordReadContext.close(LogRecordReadContext.java:399) ~[xxx]
	... 9 more
2026-07-21 12:04:14,870 INFO  org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher [] - Split fetcher 42 exited.
2026-07-21 12:04:14,870 INFO  org.apache.fluss.client.token.DefaultSecurityTokenManager    [] - Stopping security token renewal
2026-07-21 12:04:14,871 INFO  org.apache.fluss.client.token.DefaultSecurityTokenManager    [] - Stopped security token renewal
2026-07-21 12:04:16,878 INFO  org.apache.fluss.rpc.netty.client.NettyClient                [] - Netty client was shutdown successfully.
```

#### How to reproduce
1. Run a tiering job.
2. Make a lake writer fail in complete() (e.g. a transient storage error), or fail the task while lake writers are still writing.
3. The close path throws "Memory was leaked by query" instead of the real exception.

#### Root cause
TieringSplitReader never closes in-flight lake writers on failure:
- completeLakeWriter() removes the writer from the tracking map before calling complete(); if complete() throws, the writer is never closed.
- close() only closes the log scanner and table; in-flight lake writers and the snapshot split reader are left unclosed.

The leaked writers hold Arrow buffers allocated from the log scanner, so closing the
scanner afterwards fails.

#### Impact
Resources (Arrow buffers, lake write buffers, output streams) leak on every restart. The real exception is buried, making troubleshooting harder.

### Solution

Always close lake writers: on complete() failure, on split reader close (before the log scanner), and when switching tables.

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/fluss-common/src/main/java/org/apache/fluss/utils/IOUtils.java
+++ b/fluss-common/src/main/java/org/apache/fluss/utils/IOUtils.java
@@ -107,6 +107,31 @@ public class IOUtils {
         return copyBytes(in, out, BLOCKSIZE, close);
     }
 
+    /**
+     * Closes all non-null {@link AutoCloseable} objects in the parameter, suppressing exceptions.
+     * Exception will be emitted after calling close() on every object.
+     *
+     * @param closeables iterable with closeables to close.
+     * @throws Exception collected exceptions that occurred during closing
+     */
+    public static void closeAll(Iterable<? extends AutoCloseable> closeables) throws Exception {
+        if (null != closeables) {
+            Exception collectedExceptions = null;
+            for (AutoCloseable closeable : closeables) {
+                try {
+                    if (null != closeable) {
+                        closeable.close();
+                    }
+                } catch (Exception e) {
+                    collectedExceptions = ExceptionUtils.firstOrSuppressed(e, collectedExceptions);
+                }
+            }
+            if (null != collectedExceptions) {
+                throw collectedExceptions;
+            }
+        }
+    }
+
     /** Closes all elements in the iterable with closeQuietly(). */
     public static void closeAllQuietly(Iterable<? extends AutoCloseable> closeables) {
         if (null != closeables) {

--- a/fluss-common/src/test/java/org/apache/fluss/utils/IOUtilsTest.java
+++ b/fluss-common/src/test/java/org/apache/fluss/utils/IOUtilsTest.java
@@ -25,6 +25,7 @@ import java.io.IOException;
 import java.util.Arrays;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 /** Test for {@link org.apache.fluss.utils.IOUtils}. */
 class IOUtilsTest {
@@ -38,6 +39,23 @@ class IOUtilsTest {
         IOUtils.closeQuietly(withoutException);
         // test close with multi-instances
         IOUtils.closeAllQuietly(Arrays.asList(withException, withoutException));
+    }
+
+    @Test
+    void testCloseAll() throws Exception {
+        // null iterable and null elements should be tolerated
+        IOUtils.closeAll(null);
+        IOUtils.closeAll(Arrays.asList(null, new CloserWithoutException()));
+
+        // every closeable should be closed even if one fails, and the first
+        // exception should be thrown with the others suppressed
+        CloserWithException failing1 = new CloserWithException();
+        CloserWithException failing2 = new CloserWithException();
+        CloserWithoutException succeeding = new CloserWithoutException();
+        assertThatThrownBy(() -> IOUtils.closeAll(Arrays.asList(failing1, succeeding, failing2)))
+                .hasMessage("Fail to close")
+                .satisfies(t -> assertThat(t.getSuppressed()).hasSize(1));
+        assertThat(succeeding.closed).isTrue();
     }
 
     @Test
@@ -81,9 +99,13 @@ class IOUtilsTest {
     }
 
     private static class CloserWithoutException implements AutoCloseable {
+
+        private boolean closed;
+
         @Override
         public void close() throws Exception {
             // don't throw exception
+            closed = true;
         }
     }
 }

--- a/fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/tiering/source/TieringSplitReader.java
+++ b/fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/tiering/source/TieringSplitReader.java
@@ -42,6 +42,7 @@ import org.apache.fluss.metadata.TableInfo;
 import org.apache.fluss.metadata.TablePath;
 import org.apache.fluss.record.ArrowBatchData;
 import org.apache.fluss.utils.CloseableIterator;
+import org.apache.fluss.utils.IOUtils;
 import org.apache.fluss.utils.function.SupplierWithException;
 
 import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
@@ -56,6 +57,7 @@ import javax.annotation.Nullable;
 import java.io.IOException;
 import java.time.Duration;
 import java.util.ArrayDeque;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -622,7 +624,14 @@ public class TieringSplitReader<WriteResult>
         LakeWriter<WriteResult> lakeWriter = lakeWriters.remove(bucket);
         WriteResult writeResult = null;
         if (lakeWriter != null) {
-            writeResult = lakeWriter.complete();
+            try {
+                writeResult = lakeWriter.complete();
+            } catch (Exception e) {
+                // make sure the lake writer is always closed to release resources
+                // when complete fails, otherwise resources like arrow buffers leak
+                IOUtils.closeQuietly(lakeWriter, "lake writer for bucket " + bucket);
+                throw e;
+            }
             lakeWriter.close();
         }
         return toTableBucketWriteResult(
@@ -721,6 +730,10 @@ public class TieringSplitReader<WriteResult>
     }
 
     private void finishCurrentTable() throws IOException {
+        // defensive cleanup: lake writers should have been completed and closed when
+        // their splits finished, close any residual writers before closing the log
+        // scanner since their resources may be allocated from the scanner
+        closeAllLakeWriters();
         try {
             if (currentLogScanner != null) {
                 currentLogScanner.close();
@@ -776,14 +789,35 @@ public class TieringSplitReader<WriteResult>
 
     @Override
     public void close() throws Exception {
-        if (currentLogScanner != null) {
-            currentLogScanner.close();
-        }
-        if (currentTable != null) {
-            currentTable.close();
-        }
+        // lake writers must be closed before the log scanner since their resources
+        // (e.g. arrow buffers) may be allocated from the scanner;
+        // the connection is owned and closed by TieringSourceReader
+        List<AutoCloseable> closeables = new ArrayList<>(lakeWriters.values());
+        closeables.add(currentSnapshotSplitReader);
+        closeables.add(currentLogScanner);
+        closeables.add(currentTable);
 
-        // don't need to close connection, will be closed by TieringSourceReader
+        // clear the fields to make close() idempotent
+        lakeWriters.clear();
+        currentSnapshotSplitReader = null;
+        currentLogScanner = null;
+        currentTable = null;
+
+        IOUtils.closeAll(closeables);
+    }
+
+    private void closeAllLakeWriters() {
+        if (lakeWriters.isEmpty()) {
+            return;
+        }
+        LOG.warn(
+                "{} lake writer(s) of table {} are still in-flight, closing them.",
+                lakeWriters.size(),
+                currentTablePath);
+        for (Map.Entry<TableBucket, LakeWriter<WriteResult>> entry : lakeWriters.entrySet()) {
+            IOUtils.closeQuietly(entry.getValue(), "lake writer for bucket " + entry.getKey());
+        }
+        lakeWriters.clear();
     }
 
     private void subscribeLog(TieringLogSplit logSplit) {

--- a/fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/tiering/source/TieringSplitReader.java
+++ b/fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/tiering/source/TieringSplitReader.java
@@ -818,6 +818,8 @@ public class TieringSplitReader<WriteResult>
             IOUtils.closeQuietly(entry.getValue(), "lake writer for bucket " + entry.getKey());
         }
         lakeWriters.clear();
+
+        // don't need to close connection, will be closed by TieringSourceReader
     }
 
     private void subscribeLog(TieringLogSplit logSplit) {

--- a/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/tiering/TestingLakeTieringFactory.java
+++ b/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/tiering/TestingLakeTieringFactory.java
@@ -45,8 +45,20 @@ public class TestingLakeTieringFactory
 
     @Nullable private TestingLakeCommitter testingLakeCommitter;
 
+    // the exception to be thrown when complete() is called on created lake writers
+    @Nullable private final IOException writerCompleteException;
+
+    private final List<TestingLakeWriter> createdLakeWriters = new ArrayList<>();
+
     public TestingLakeTieringFactory(@Nullable TestingLakeCommitter testingLakeCommitter) {
+        this(testingLakeCommitter, null);
+    }
+
+    public TestingLakeTieringFactory(
+            @Nullable TestingLakeCommitter testingLakeCommitter,
+            @Nullable IOException writerCompleteException) {
         this.testingLakeCommitter = testingLakeCommitter;
+        this.writerCompleteException = writerCompleteException;
     }
 
     public TestingLakeTieringFactory() {
@@ -59,7 +71,13 @@ public class TestingLakeTieringFactory
     @Override
     public LakeWriter<TestingWriteResult> createLakeWriter(WriterInitContext writerInitContext)
             throws IOException {
-        return new TestingLakeWriter();
+        TestingLakeWriter lakeWriter = new TestingLakeWriter(writerCompleteException);
+        createdLakeWriters.add(lakeWriter);
+        return lakeWriter;
+    }
+
+    public List<TestingLakeWriter> getCreatedLakeWriters() {
+        return createdLakeWriters;
     }
 
     @Override
@@ -82,9 +100,22 @@ public class TestingLakeTieringFactory
                 "method getCommittableSerializer is not supported.");
     }
 
-    private static final class TestingLakeWriter implements LakeWriter<TestingWriteResult> {
+    /** A lake writer for testing purpose which tracks the closed state. */
+    public static final class TestingLakeWriter implements LakeWriter<TestingWriteResult> {
 
         private int writtenRecords;
+
+        @Nullable private final IOException completeException;
+
+        private boolean closed;
+
+        public TestingLakeWriter() {
+            this(null);
+        }
+
+        public TestingLakeWriter(@Nullable IOException completeException) {
+            this.completeException = completeException;
+        }
 
         @Override
         public void write(LogRecord record) throws IOException {
@@ -93,11 +124,20 @@ public class TestingLakeTieringFactory
 
         @Override
         public TestingWriteResult complete() throws IOException {
+            if (completeException != null) {
+                throw completeException;
+            }
             return new TestingWriteResult(writtenRecords);
         }
 
         @Override
-        public void close() throws IOException {}
+        public void close() throws IOException {
+            closed = true;
+        }
+
+        public boolean isClosed() {
+            return closed;
+        }
     }
 
     /** A lake committer for testing purpose. */

--- a/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/tiering/source/TieringSplitReaderTest.java
+++ b/fluss-flink/fluss-flink-common/src/test/java/org/apache/fluss/flink/tiering/source/TieringSplitReaderTest.java
@@ -65,6 +65,7 @@ import java.util.stream.Collectors;
 import static org.apache.fluss.client.table.scanner.log.LogScanner.EARLIEST_OFFSET;
 import static org.apache.fluss.testutils.DataTestUtils.row;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 /** UT for {@link TieringSplitReader}. */
 class TieringSplitReaderTest extends FlinkTestBase {
@@ -73,11 +74,12 @@ class TieringSplitReaderTest extends FlinkTestBase {
     void testTieringTable() throws Exception {
         TablePath tablePath = TablePath.of("fluss", "fluss_test_tiering_one_table");
         long tableId = createTable(tablePath, DEFAULT_PK_TABLE_DESCRIPTOR);
+        TestingLakeTieringFactory lakeTieringFactory = new TestingLakeTieringFactory();
         try (Connection connection =
                         ConnectionFactory.createConnection(
                                 FLUSS_CLUSTER_EXTENSION.getClientConfig());
                 TieringSplitReader<TestingWriteResult> tieringSplitReader =
-                        createTieringReader(connection)) {
+                        createTieringReader(connection, lakeTieringFactory)) {
             // test empty splits
             SplitsAddition<TieringSplit> splitsAddition =
                     new SplitsAddition<>(
@@ -161,6 +163,11 @@ class TieringSplitReaderTest extends FlinkTestBase {
             tieringSplitReader.handleSplitsChanges(new SplitsAddition<>(logSplits));
             verifyTieringRows(
                     tieringSplitReader, tableId, expectedRowCount, expectFinishTieringSplits);
+
+            // all created lake writers must be completed and closed with the splits
+            assertThat(lakeTieringFactory.getCreatedLakeWriters()).isNotEmpty();
+            assertThat(lakeTieringFactory.getCreatedLakeWriters())
+                    .allSatisfy(writer -> assertThat(writer.isClosed()).isTrue());
         }
     }
 
@@ -335,6 +342,105 @@ class TieringSplitReaderTest extends FlinkTestBase {
             assertThat(writeResult).isNotNull();
             // expect null write result since no any records written
             assertThat(writeResult.writeResult()).isNull();
+        }
+    }
+
+    /**
+     * Verifies the lake writer is always closed even if {@link LakeWriter#complete()} fails,
+     * otherwise the writer is leaked since it has been removed from the tracking map and can never
+     * be reached again.
+     */
+    @Test
+    void testLakeWriterClosedWhenCompleteFails() throws Exception {
+        TablePath tablePath = TablePath.of("fluss", "tiering_close_writer_on_complete_failure");
+        TableDescriptor singleBucketPkTableDescriptor =
+                TableDescriptor.builder()
+                        .schema(DEFAULT_PK_TABLE_SCHEMA)
+                        .distributedBy(1, "id")
+                        .build();
+        long tableId = createTable(tablePath, singleBucketPkTableDescriptor);
+        // the factory creates lake writers whose complete() always throws the given exception
+        TestingLakeTieringFactory lakeTieringFactory =
+                new TestingLakeTieringFactory(null, new IOException("Injected complete failure"));
+        try (Connection connection =
+                        ConnectionFactory.createConnection(
+                                FLUSS_CLUSTER_EXTENSION.getClientConfig());
+                Table table = connection.getTable(tablePath);
+                TieringSplitReader<TestingWriteResult> tieringSplitReader =
+                        createTieringReader(connection, lakeTieringFactory)) {
+            // write 2 records which occupy log offset 0 and 1
+            UpsertWriter upsertWriter = table.newUpsert().createWriter();
+            upsertWriter.upsert(row(1, "v1"));
+            upsertWriter.upsert(row(2, "v2"));
+            upsertWriter.flush();
+
+            // add a log split covering all the written records, so that once the split reaches
+            // the stopping offset (2), the reader completes the lake writer
+            TableBucket tableBucket = new TableBucket(tableId, 0);
+            tieringSplitReader.handleSplitsChanges(
+                    new SplitsAddition<>(
+                            Collections.singletonList(
+                                    new TieringLogSplit(
+                                            tablePath, tableBucket, null, EARLIEST_OFFSET, 2, 1))));
+
+            // the injected complete() failure should propagate out of fetch()
+            assertThatThrownBy(
+                            () -> {
+                                for (int i = 0; i < 10; i++) {
+                                    tieringSplitReader.fetch();
+                                }
+                            })
+                    .hasMessageContaining("Injected complete failure");
+
+            // the writer must be closed although complete() failed
+            assertThat(lakeTieringFactory.getCreatedLakeWriters()).hasSize(1);
+            assertThat(lakeTieringFactory.getCreatedLakeWriters().get(0).isClosed()).isTrue();
+        }
+    }
+
+    /**
+     * Verifies {@link TieringSplitReader#close()} closes all in-flight lake writers, which is the
+     * case when the split reader is closed on task failure before the splits finish.
+     */
+    @Test
+    void testCloseClosesAllInFlightLakeWriters() throws Exception {
+        TablePath tablePath = TablePath.of("fluss", "tiering_close_in_flight_writers");
+        long tableId = createTable(tablePath, DEFAULT_PK_TABLE_DESCRIPTOR);
+        TestingLakeTieringFactory lakeTieringFactory = new TestingLakeTieringFactory();
+        try (Connection connection =
+                        ConnectionFactory.createConnection(
+                                FLUSS_CLUSTER_EXTENSION.getClientConfig());
+                TieringSplitReader<TestingWriteResult> tieringSplitReader =
+                        createTieringReader(connection, lakeTieringFactory)) {
+            Map<TableBucket, List<InternalRow>> rows = putRows(tableId, tablePath, 10);
+
+            // add log splits with a stopping offset beyond the log end offset, so the splits
+            // never finish and the lake writers stay in-flight
+            List<TieringSplit> logSplits = new ArrayList<>();
+            for (Map.Entry<TableBucket, List<InternalRow>> entry : rows.entrySet()) {
+                logSplits.add(
+                        createLogSplit(
+                                tablePath,
+                                tableId,
+                                entry.getKey().getBucket(),
+                                EARLIEST_OFFSET,
+                                entry.getValue().size() + 100));
+            }
+            tieringSplitReader.handleSplitsChanges(new SplitsAddition<>(logSplits));
+
+            // fetch until a lake writer has been created for every bucket with records
+            for (int i = 0;
+                    i < 10 && lakeTieringFactory.getCreatedLakeWriters().size() < rows.size();
+                    i++) {
+                tieringSplitReader.fetch();
+            }
+            assertThat(lakeTieringFactory.getCreatedLakeWriters()).hasSize(rows.size());
+
+            // close the reader while all the writers are still in-flight,
+            // all of them must be closed to avoid resource leaks
+            tieringSplitReader.close();
+            assertThat(lakeTieringFactory.getCreatedLakeWriters())
+                    .allSatisfy(writer -> assertThat(writer.isClosed()).isTrue());
         }
     }
 

